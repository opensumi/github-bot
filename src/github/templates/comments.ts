import {
  renderPrOrIssueLink,
  renderPrOrIssueText,
  renderRepoLink,
  renderUserLink,
} from '.';
import { ExtractPayload } from '../template';

export function handleIssueComment(payload: ExtractPayload<'issue_comment'>) {
  const issue = payload.issue;
  const action = payload.action;
  const comment = payload.comment;
  const isUnderPullRequest = Boolean(issue.pull_request);
  let location = 'issue';
  if (isUnderPullRequest) {
    location = 'pull request';
  }
  const title = `Comment ${action} on ${location} ${renderPrOrIssueText(
    payload.issue,
  )}`;
  const text = `${renderRepoLink(payload.repository)} [Comment](${
    comment.html_url
  }) ${action} by ${renderUserLink(
    payload.sender,
  )} on ${location} ${renderPrOrIssueLink(payload.issue)}
>
> ${comment.body}
`;
  return {
    title,
    text,
  };
}
