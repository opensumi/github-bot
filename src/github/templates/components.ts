import capitalize from 'lodash/capitalize';

import { StringBuilder, limitTextByPosition } from '@/utils/string-builder';

import { render } from '../renderer';

export function RepositoryLink(repository: { name: string; html_url: string }) {
  return `[[${escapeUsername(repository.name)}]](${repository.html_url})`;
}

function escapeUsername(text: string) {
  return text.replace(/\[/g, '\\[').replace(/\]/g, '\\]');
}

export function SenderLink(sender: { login: string; html_url: string }) {
  return `[${escapeUsername(sender.login)}](${sender.html_url})`;
}

export function TeamLink(team: { name: string; html_url: string }) {
  return `[${escapeUsername(team.name)}](${team.html_url})`;
}

export function AtSenderLink(sender: { login: string; html_url: string }) {
  return `[@${escapeUsername(sender.login)}](${sender.html_url})`;
}

export function ReleaseLink(release: {
  name: string;
  tag_name: string;
  html_url: string;
}) {
  return `[${escapeUsername(release.name || release.tag_name)}](${
    release.html_url
  })`;
}

export function IssuesText(
  p: {
    /**
     * The title of the pull request.
     */
    title: string;
    /**
     * Number uniquely identifying the pull request within its repository.
     */
    number?: number;
  },
  prefix?: string,
  suffix?: string,
) {
  let text = `${prefix ?? ''}`;
  if (p.number) {
    text += `#${p.number} `;
  }
  text += p.title;
  if (suffix) {
    text += suffix;
  }
  return text;
}

export function IssuesLink(
  p: {
    /**
     * The title of the pull request.
     */
    title: string;
    /**
     * Number uniquely identifying the pull request within its repository.
     */
    number?: number;
    html_url: string;
  },
  prefix?: string,
  suffix?: string,
) {
  return `[${IssuesText(p, prefix, suffix)}](${p.html_url})`;
}

export function IssuesTitleLink(p: {
  /**
   * The title of the pull request.
   */
  title: string;
  /**
   * Number uniquely identifying the pull request within its repository.
   */
  number: number;
  html_url: string;
  body?: string | null | undefined;
}) {
  return `#### ${IssuesLink(p)}`;
}

export function PullRequestRefInfo(data: {
  head: {
    label: string;
  };
  base: {
    ref: string;
    user: {
      login: string;
    };
  };
}) {
  // display PR related info, such as pr assignees, base branch, head branch, etc.
  const head = data.head;
  const base = data.base;
  const headLabel = removeOrgInfo(base.user.login, head.label);
  return `${base.ref} <- ${headLabel}  `;
}

export function AssigneesInfo(
  assignees: {
    login: string;
    html_url: string;
  }[],
) {
  const assigneeNames = assignees.map((v) => SenderLink(v)).join(', ');
  return `Assignees: ${assigneeNames}  `;
}

interface ISenderBasic {
  login: string;
  html_url: string;
}

interface ITeamBasic {
  name: string;
  html_url: string;
}

export function RequestedReviewersInfo(
  reviewers: (ISenderBasic | ITeamBasic)[],
) {
  const reviewerNames = reviewers
    .map((v) =>
      (v as ISenderBasic).login
        ? SenderLink(v as ISenderBasic)
        : TeamLink(v as ITeamBasic),
    )
    .join(', ');
  return `Requested reviewers: ${reviewerNames}  `;
}

export function DeletedPrOrIssueTitleLink(p: {
  /**
   * The title of the pull request.
   */
  title: string;
  /**
   * Number uniquely identifying the pull request within its repository.
   */
  number: number;
  html_url: string;
  body: string | null | undefined;
}) {
  return `#### ${IssuesLink(p, '~~', '~~')}`;
}

export function Reference(text?: string | null | undefined, bodyLimit = -1) {
  if (!text) {
    return '';
  }

  if (bodyLimit && bodyLimit > 0) {
    text = limitTextByPosition(text, bodyLimit);
  }

  const arrayOfLines = text.replace(/\r\n|\n\r|\n|\r/g, '\n').split('\n');
  const newLines = [];

  for (const line of arrayOfLines) {
    if (line.trim()) {
      newLines.push(`> ${line}`);
    } else {
      newLines.push(`>`);
    }
  }

  return newLines.join('\n');
}

export class StopHandleError extends Error {
  constructor(reason: string) {
    super(reason);
  }
}

export type TextTplInput = {
  payload: any;
  title: string;
  target?: string;
  compactTitle?: string;
  body?: string | null;
  event: string;
  action?: string;
  contentLimit?: number;
  notCapitalizeTitle?: boolean;
  doNotRenderBody?: boolean;
  autoRef?: boolean;

  blockUsers?: Set<string>;
  blockActions?: Set<string>;
};

export type TemplateRenderResult = {
  title: string;
  text: string;
  compactText?: string;
};

type TextTpl = (
  data: TextTplInput,
  ctx?: {
    setting?: {
      notDisplayRepoName?: boolean;
    };
  },
) => TemplateRenderResult;

export type HandlerResult = {
  text: TextTplInput;
};

export const depManageBotToIgnore = new Set([
  'renovate[bot]',
  'dependabot[bot]',
]);

export const Template: TextTpl = (data, ctx) => {
  const {
    payload,
    title: bodyTitle,
    compactTitle,
    target,
    body,
    notCapitalizeTitle,
    autoRef,
    contentLimit = -1,
    blockActions,
    event,
    blockUsers,
  } = data;
  const action = data.action ?? payload.action;

  if (!action) {
    throw new Error('action is required for event ' + event);
  }

  if (blockActions && blockActions.has(action)) {
    throw new StopHandleError(
      `skip event ${event} action ${action} because it is in blockActions`,
    );
  }

  if (blockUsers && payload.sender && blockUsers.has(payload.sender.login)) {
    throw new StopHandleError(
      `skip event ${event} action ${action} because it is in blockUsers`,
    );
  }

  const repo = payload.repository;

  let repoInfo = RepositoryLink(repo) + ' ';
  if (ctx?.setting?.notDisplayRepoName) {
    repoInfo = '';
  }

  const textBuilder = new StringBuilder(
    `#### ${repoInfo}${bodyTitle.trim()}  `,
  );
  let compactTextBuilder: StringBuilder | undefined;
  if (compactTitle) {
    compactTextBuilder = new StringBuilder(compactTitle);
  }

  if (target) {
    textBuilder.add(target);
    // compact text do not need target
  }

  let bodyText = '';
  if (body && !data.doNotRenderBody) {
    bodyText = render(body.trim(), payload);
  }

  if (bodyText) {
    textBuilder.addDivider('', true);

    if (contentLimit && contentLimit > 0) {
      bodyText = limitTextByPosition(bodyText, contentLimit);
    }

    payload.bodyText = bodyText;

    const refText = autoRef ? '{{bodyText|ref}}' : '{{bodyText}}';
    textBuilder.add(refText);
    compactTextBuilder && compactTextBuilder.add(refText);
  }

  let titleText = `${notCapitalizeTitle ? event : capitalize(event)} ${action}`;
  if (!ctx?.setting?.notDisplayRepoName) {
    titleText = `[{{repository.name}}] ${titleText}`;
  }

  const titleBuilder = new StringBuilder(titleText);

  return {
    title: titleBuilder.render(payload),
    text: textBuilder.render(payload),
    compactText: compactTextBuilder
      ? compactTextBuilder.render(payload)
      : undefined,
  };
};

const removeOrgInfo = (orgName: string, label: string) => {
  const prefix = `${orgName}:`;
  if (label.startsWith(prefix)) {
    return label.slice(prefix.length);
  }
  return label;
};

/**
 * `not_planned` -> `not planned`
 */
export function prettyUnderlineWord(stateReason: string) {
  return stateReason.replace(/_/g, ' ');
}
