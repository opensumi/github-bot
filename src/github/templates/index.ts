export * from './prOrIssue';
export * from './comment';
export * from './release';
export * from './review';
export * from './utils';

import {
  handlePr,
  handleIssue,
  handleReview,
  handleRelease,
  handleDiscussion,
  handleIssueComment,
  handleDiscussionComment,
} from '.';

import { EmitterWebhookEventName } from '@octokit/webhooks/dist-types/types';
import { TemplateMapping } from '@/github/types';

export const templates = {
  'issues.opened': handleIssue,
  'issues.closed': handleIssue,
  'issues.reopened': handleIssue,
  'issues.edited': handleIssue,
  'pull_request.opened': handlePr,
  'pull_request.reopened': handlePr,
  'pull_request.closed': handlePr,
  'pull_request.edited': handlePr,
  'pull_request.ready_for_review': handlePr,
  'discussion.created': handleDiscussion,
  'discussion.edited': handleDiscussion,
  'discussion.deleted': handleDiscussion,
  'discussion_comment.created': handleDiscussionComment,
  'discussion_comment.deleted': handleDiscussionComment,
  'issue_comment.created': handleIssueComment,
  'issue_comment.deleted': handleIssueComment,
  'release.published': handleRelease,
  'release.released': handleRelease,
  'pull_request_review.submitted': handleReview,
  'pull_request_review.edited': handleReview,
  'pull_request_review.dismissed': handleReview,
} as TemplateMapping;

export const supportTemplates = Object.keys(
  templates,
) as EmitterWebhookEventName[];
