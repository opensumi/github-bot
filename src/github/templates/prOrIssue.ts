import { ExtractPayload } from '../types';
import { renderRepoLink, renderUserLink, renderPrOrIssue } from './utils';
import { Issue, PullRequest, Discussion } from '@octokit/webhooks-types';
import { StringBuilder } from '@/utils';
import { titleTpl } from './trivias';
import { Context } from '../app';

type Name = 'issues' | 'pull_request' | 'discussion';
const NameBlock = {
  issues: 'issue',
  pull_request: 'pull request',
  discussion: 'discussion',
} as {
  [key in Name]: string;
};

function render(
  name: Name,
  payload: ExtractPayload<Name>,
  data: PullRequest | Issue | Discussion,
  ctx: Context,
) {
  const nameBlock = NameBlock[name];
  let action = payload.action.replaceAll('_', '');

  let shouldRenderBody = true;
  if (['closed', 'edited'].includes(action)) {
    shouldRenderBody = false;
  }

  const subline = [] as string[];

  if (name === 'pull_request' && action === 'closed') {
    const pr = (payload as ExtractPayload<'pull_request.closed'>).pull_request;
    if (pr.merged) {
      // If the action is closed and the merged key is true, the pull request was merged.
      action = 'merged';
    }
  }

  const title = titleTpl({
    repo: payload.repository,
    event: `${nameBlock}#${data.number}`,
    action,
  });

  const builder = new StringBuilder(
    `#### ${renderRepoLink(payload.repository)} ${renderUserLink(
      payload.sender,
    )} ${action} [${nameBlock}#${data.number}](${data.html_url})`,
  );

  if (subline.length > 0) {
    builder.add(subline.join(', '), true);
  }
  builder.add(
    renderPrOrIssue(data, shouldRenderBody, ctx.dingSecret.contentLimit),
  );

  return {
    title,
    text: builder.build(),
  };
}

export async function handlePr(
  payload: ExtractPayload<'pull_request'>,
  ctx: Context,
) {
  return render('pull_request', payload, payload.pull_request, ctx);
}

export async function handleIssue(
  payload: ExtractPayload<'issues'>,
  ctx: Context,
) {
  return render('issues', payload, payload.issue, ctx);
}

export async function handleDiscussion(
  payload: ExtractPayload<'discussion'>,
  ctx: Context,
) {
  return render('discussion', payload, payload.discussion, ctx);
}
