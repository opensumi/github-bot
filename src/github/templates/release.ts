import { StringBuilder } from '@/utils';

import { Context , ExtractPayload } from '../types';

import { titleTpl } from './utils';

import { renderUserLink, useRef, textTpl } from '.';

export async function handleRelease(
  payload: ExtractPayload<'release'>,
  ctx: Context,
) {
  const action = payload.action;
  const release = payload.release;
  const repo = payload.repository;
  const title = titleTpl(
    {
      repo,
      event: release.name,
      action,
    },
    ctx,
    false,
  );

  const builder = new StringBuilder();

  let status = '';
  if (release.draft) {
    status = 'Draft';
  }
  if (release.prerelease) {
    status = 'Pre Release';
  }

  if (status) {
    builder.add(`Status: ${status}\n`);
  }

  builder.add(`Tag: ${release.tag_name}\n`);
  builder.add(`>\n${useRef(release.body)}`);

  const text = textTpl(
    {
      repo: payload.repository,
      title: `${renderUserLink(payload.sender)} ${action} [${release.name}](${
        release.html_url
      })`,
      body: builder.build(),
    },
    ctx,
  );

  return { title, text };
}
