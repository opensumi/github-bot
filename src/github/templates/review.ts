import { StringBuilder } from '@/utils';
import {
  renderUserLink,
  StopHandleError,
  useRef,
  titleTpl,
  renderPrOrIssue,
  textTpl,
} from '.';
import { Context } from '../app';
import { ExtractPayload, MarkdownContent } from '../types';

export async function handleReview(
  payload: ExtractPayload<'pull_request_review'>,
  ctx: Context,
): Promise<MarkdownContent> {
  const review = payload.review;
  const action = payload.action as string;
  const pr = payload.pull_request;

  if (action === 'submitted' && review.state === 'commented') {
    throw new StopHandleError(
      'review comment is handled by handleReviewComment',
    );
  }

  let titleActionText = action;
  let fullAction = action;

  if (review.state) {
    titleActionText = review.state as string;
  }

  if (action === 'changes_requested') {
    titleActionText = 'requested changes';
    fullAction = 'changes requested';
  }

  if (action === 'dismissed') {
    fullAction = 'dismissed their stale review';
  }

  const title = titleTpl(
    {
      repo: payload.repository,
      event: 'review',
      action: titleActionText,
    },
    ctx,
  );

  const builder = new StringBuilder();
  builder.add(renderPrOrIssue(pr, false));
  builder.add(useRef(review.body, ctx.setting.contentLimit));

  const text = textTpl(
    {
      title: `${renderUserLink(payload.sender)} [${fullAction}](${
        review.html_url
      }) on [pull request](${pr.html_url})`,
      body: builder.build(),
      repo: payload.repository,
    },
    ctx,
  );

  return {
    title,
    text,
  };
}
