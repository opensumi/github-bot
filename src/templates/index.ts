export * from './prOrIssue';
export * from './utils';
