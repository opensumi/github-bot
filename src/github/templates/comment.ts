import {
  limitLine,
  renderPrOrIssueLink,
  renderPrOrIssueText,
  renderRepoLink,
  renderUserLink,
  useRef,
} from '.';
import { ExtractPayload } from '@/github/types';
import { Issue, PullRequest, Discussion, User } from '@octokit/webhooks-types';

type Name = 'issues' | 'pull_request' | 'discussion';
const NameBlock = {
  issues: 'issue',
  pull_request: 'pull request',
  discussion: 'discussion',
} as {
  [key in Name]: string;
};

const formatByUserLogin = {
  'codecov-commenter': (text: string) => {
    return limitLine(text, 3);
  },
} as {
  [key: string]: (text: string) => string;
};

function renderCommentBody(comment: { body: string; user: User }) {
  let text = useRef(comment.body);
  const formatter = formatByUserLogin[comment.user.login];
  if (formatter) {
    text = formatter(text);
  }
  return text;
}

function renderComment(
  name: Name,
  payload: ExtractPayload<'issue_comment' | 'discussion_comment'>,
  data: Issue | PullRequest | Discussion,
) {
  const comment = payload.comment;
  const location = NameBlock[name];
  const action = payload.action;
  let shouldRenderBody = true;
  if (['edited'].includes(action)) {
    shouldRenderBody = false;
  }

  const title = `[${
    payload.repository.name
  }] comment ${action} on ${location} ${renderPrOrIssueText(data)}`;
  const text = `${renderRepoLink(payload.repository)} ${renderUserLink(
    payload.sender,
  )} ${action} [comment](${
    comment.html_url
  }) on ${location} ${renderPrOrIssueLink(data)}${
    shouldRenderBody ? `\n>\n${renderCommentBody(payload.comment)}` : ''
  }
`;
  return {
    title,
    text,
  };
}

export function handleIssueComment(payload: ExtractPayload<'issue_comment'>) {
  const issue = payload.issue;
  const isUnderPullRequest = Boolean(issue.pull_request);
  let name = 'issues' as Name;
  if (isUnderPullRequest) {
    name = 'pull_request';
  }
  return renderComment(name, payload, issue);
}

export function handleDiscussionComment(
  payload: ExtractPayload<'discussion_comment'>,
) {
  return renderComment('discussion', payload, payload.discussion);
}

export function handleCommitComment(payload: ExtractPayload<'commit_comment'>) {
  const comment = payload.comment;
  const commitId = comment.commit_id.slice(0, 6);
  const action = payload.action;

  const title = `[${payload.repository.name}] commit comment ${action} on commit@${commitId}`;
  const text = `${renderRepoLink(payload.repository)} ${renderUserLink(
    payload.sender,
  )} ${action} [commit comment](${comment.html_url}) on [${
    payload.repository.name
  }@${commitId}](${comment.html_url})
>\n
${renderCommentBody(payload.comment)}
`;
  return {
    title,
    text,
  };
}
