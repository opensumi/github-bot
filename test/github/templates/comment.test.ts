import { handleCommitComment } from '../../../src/github/templates/comment';
import { commit_comment_611372_0_created } from '../../fixtures';
import { ctx } from '../ctx';

describe('github templates comment', () => {
  it('can handle commit_comment', async () => {
    const result = await handleCommitComment(
      commit_comment_611372_0_created,
      ctx,
    );
    console.log(`commit_comment ~ result`, result);
    expect(result).toMatchSnapshot();
    expect(result.text).toBeDefined();
    expect(result.title).toBeDefined();
  });
});
