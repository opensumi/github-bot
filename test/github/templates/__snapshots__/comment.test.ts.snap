// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github templates comment can handle commit_comment 1`] = `
{
  "compactText": undefined,
  "text": "#### [Codertocat](https://github.com/Codertocat) created comment on [commit@611372](https://github.com/Codertocat/Hello-World/commit/6113728f27ae82c7b1a177c8d03f9e96e0adf246#commitcomment-33548674)  
---
> > #### [M: ](https://github.com/Codertocat/Hello-World/commit/6113728f27ae82c7b1a177c8d03f9e96e0adf246#commitcomment-33548674)
> >
> > This is a really good change! :+1:",
  "title": "Commit comment created",
}
`;
